In [ ]:
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

# Початкові параметри
baseline_conversion = 0.10
relative_growth_3 = 0.03
alpha_5 = 0.05
power = 0.80

# Очікувана конверсія у тестовій групі
conversion_3 = baseline_conversion * (1 + relative_growth_3)

# Розрахунок effect size
effect_size_3 = proportion_effectsize(
    baseline_conversion,
    conversion_3
)

# Розрахунок необхідного розміру вибірки
analysis = NormalIndPower()

sample_size_3_alpha5 = analysis.solve_power(
    effect_size=effect_size_3,
    power=power,
    alpha=alpha_5,
    ratio=1
)

print(f"Базова конверсія: {baseline_conversion:.1%}")
print(f"Відносний ріст: {relative_growth_3:.1%}")
print(f"Очікувана конверсія: {conversion_3:.1%}")
print(f"Effect size: {effect_size_3:.4f}")
print(f"Розмір вибірки на одну групу: {sample_size_3_alpha5:.0f}")
print(f"Загальний розмір вибірки: {sample_size_3_alpha5 * 2:.0f}")

Базова конверсія: 10.0%
Відносний ріст: 3.0%
Очікувана конверсія: 10.3%
Effect size: -0.0099
Розмір вибірки на одну групу: 159058
Загальний розмір вибірки: 318116


За базової конверсії 10%, очікуваного відносного росту на 3%, рівня значущості α = 5% та статистичної потужності 80% очікувана конверсія в тестовій групі становить 10.3%. Для виявлення такого ефекту необхідно приблизно 159 000 користувачів у кожній групі, тобто близько 318 000 користувачів загалом

In [ ]:
alpha_1 = 0.01

sample_size_alpha1 = analysis.solve_power(
    effect_size=effect_size_3,
    power=power,
    alpha=alpha_1,
    ratio=1
)

print(f"Рівень значущості: {alpha_1:.0%}")
print(f"Розмір вибірки на одну групу: {sample_size_alpha1:.0f}")
print(f"Загальний розмір вибірки: {sample_size_alpha1 * 2:.0f}")

Рівень значущості: 1%
Розмір вибірки на одну групу: 236676
Загальний розмір вибірки: 473352


При зниженні рівня значущості з 5% до 1% необхідний розмір вибірки збільшується, оскільки для підтвердження статистично значущого результату потрібні сильніші докази. За тієї самої базової конверсії 10%, відносного росту 3% та потужності 80% вибірка має бути більшою, ніж при α = 5%.

In [ ]:
relative_growth_1 = 0.01

# Очікувана конверсія при відносному рості 1%
conversion_1 = baseline_conversion * (1 + relative_growth_1)

# Розрахунок effect size
effect_size_1 = proportion_effectsize(
    baseline_conversion,
    conversion_1
)

# Розрахунок необхідного розміру вибірки
sample_size_effect1 = analysis.solve_power(
    effect_size=effect_size_1,
    power=power,
    alpha=alpha_5,
    ratio=1
)

print(f"Базова конверсія: {baseline_conversion:.1%}")
print(f"Відносний ріст: {relative_growth_1:.1%}")
print(f"Очікувана конверсія: {conversion_1:.1%}")
print(f"Effect size: {effect_size_1:.4f}")
print(f"Розмір вибірки на одну групу: {sample_size_effect1:.0f}")
print(f"Загальний розмір вибірки: {sample_size_effect1 * 2:.0f}")

Базова конверсія: 10.0%
Відносний ріст: 1.0%
Очікувана конверсія: 10.1%
Effect size: -0.0033
Розмір вибірки на одну групу: 1419062
Загальний розмір вибірки: 2838124


Якщо очікуваний відносний ріст конверсії зменшується з 3% до 1%, очікувана конверсія змінюється з 10% до 10.1%. Оскільки ефект стає значно меншим, для його виявлення потрібна суттєво більша вибірка. Отже, чим менший очікуваний ефект, тим більший розмір вибірки необхідний для A/B-тесту.